onnxruntime/core/mlas/lib/aarch64/DepthwiseQConvKernelSize9Neon.S

```diff
@@ -0,0 +1,366 @@
+        ins     v12.d[0], x1                // Filter
+        ins     v13.d[0], x9                // Bias
+        ins     v13.d[1], x8                // Scale
+        ld1r    {v0.8h}, [x5]               // zero point
+        dup     v5.4s, w10                  // flip 0x80
+
+        tbnz    x6, #.LMLAS_CONV_SYM_FLAG_PER_CHANNEL_SCALE_BIT_INDEX, .LMlasConvSymDepthwiseKernelSize9_SkipPerTensorScaleInit
+        ld1r    {v1.4s}, [x8]               // load scale value
+        mov     v2.16b, v1.16b
+        mov     v3.16b, v1.16b
+        mov     v4.16b, v1.16b
+
+.LMlasConvSymDepthwiseKernelSize9_SkipPerTensorScaleInit:
+
+        add     x9, x3, x2                  // x9 <---- Ouput1
+        cbz     x4, .LMlasConvSymDepthwiseKernelSize9_Exit
+
+.LMlasConvSymDepthwiseKernelSize9_OutputLoop:
+        ldp     x20, x21, [x0], #72         // input ptrs for Output0
+        ldp     x22, x23, [x0, #-56]
+        sub     x4, x4, #1
+        ldp     x24, x25, [x0, #-40]
+        ldp     x26, x27, [x0, #-24]
+        ldur    x28, [x0, #-8]
+        
+        cbz     x4, .LMlasConvSymDepthwiseKernelSize9_Dup_Inputs
+        ldp     x10, x11, [x0], #72         // input ptrs for Output0
+        ldp     x12, x13, [x0, #-56]
+        sub     x4, x4, #1
+        ldp     x14, x15, [x0, #-40]
+        ldp     x16, x17, [x0, #-24]
+        ldur    x19, [x0, #-8]
+        b       .LMlasConvSymDepthwiseKernelSize9_Loaded_Input
+
+.LMlasConvSymDepthwiseKernelSize9_Dup_Inputs:
+        mov     x9, x3                      // Output1 <-- Output0
+        mov     x10, x20
+        mov     x11, x21
+        mov     x12, x22
+        mov     x13, x23
+        mov     x14, x24
+        mov     x15, x25
+        mov     x16, x26
+        mov     x17, x27
+        mov     x19, x28
+
+.LMlasConvSymDepthwiseKernelSize9_Loaded_Input:
+
+        eor     x8, x8, x8                  // Processed channels
+        umov    x1, v12.D[0]                // filter
+        umov    x5, v13.D[0]                // bias
+        umov    x7, v13.D[1]                // scale
+    
+        cmp     x8, x2                      // Save one register by not using count down to zero here
+        bhs     .LMlasConvSymDepthwiseKernelSize9_Finish_Channels16_Loop
+
+.LMlasConvSymDepthwiseKernelSize9_Channels16_Loop:
+        ld1     {v10.16b}, [x1], x2         // vk0
+        ldr     q16, [x20, x8]              // out0 vi0
+        ldr     q17, [x10, x8]              // out1 vi0
+        ld1     {v6.4s, v7.4s, v8.4s, v9.4s}, [x5], #64       // bias vacc 0-15 for outs
+        ld1     {v11.16b}, [x1], x2         // vk1
+        ldr     q18, [x21, x8]              // out0 vi1
+        ldr     q19, [x11, x8]              // out1 vi1
+
+        eor     v16.16b, v16.16b, v5.16b    // -128 to signed int8
+        eor     v17.16b, v17.16b, v5.16b
+        eor     v18.16b, v18.16b, v5.16b
+        eor     v19.16b, v19.16b, v5.16b
+
+        smull   v24.8h, v10.8b, v16.8b
```

Comment:
The use of vector registers here is quite close to their corresponding load instructions. In the case where the data is not in l1 cache, there maybe pipeline bubble here. Maybe try moving the next batch of load instructions (line 171 - 176) before the 4 eor above. This way you reduce the chance of waiting for all these loads and maybe we can save tens of cycles per iteration.

Reply:
Nice suggestion.